design/pool-configuration.md

```diff
@@ -0,0 +1,552 @@
+into the CRDs, or provide some offline tool in charge of that (the latter option is more desirable, as it will remove the need of reconciling changes into the configmap).
+
+#### Impacts on users currently using the MetalLB operator
+
+The operator already exposes CRDs, mapped directly to the current configmap entitites.
+Changing the CRDs layout will impact them, and in order to mitigate the behaviour, conversion webhooks will be implemented to support both the old and the new versions
+of the API, until the old version is deprecated.
+
+#### Bad configurations
+
+##### Service allocation
+
+When a user specify both a pool to be allocated to a specific service and one with no selector, the service may receive an IP from the broader one.
+We need to implement an easy to understand and deterministic behaviour (see design details below).
+
+##### Non overlapping filters
+
+Having an IP with no nodes advertising it is a risk. Some scenarios are:
+
+* Traffic Policy Local with pods not running on the same nodes selected by the selector
+* Node BGP peer selector with no intersection with the address selector
+
+It should be possible to detect non overlapping configurations, and raise an event when they happen. The metallb operator introduces a Validation WebHook
+that can be used to prevent those configurations to be created in first instance.
+
+### Conflicts with discovery
+
+The peer discovery feature currently being implemented in [this PR](https://github.com/metallb/metallb/pull/593) will introduce a way to
+fetch peers from the node's annotations.
+
+With this new structure, the PR will need to be adjusted to extend the `BGPPeer` entity in order to consume the node's annotations instead of hardcoded values.
+
+### Conflicts with the FRR implementation
+
+All the BGP changes will happen from a "behavioural" point of view. Where needed, the changes will reflected in both the native and the new implementation.
+
+### Not covering all the angles
+
+The risk of not covering all possible cases / permutations is still there.
+
+That aside, this proposal covers the majority of the issues that were raised by the users (multiple times), and it won't prevent to add a finer level of configuration
+if ever comes the need. Moreover, providing a straightforward way to separate concerns will make it easier to accomodate new changes, and it will be more clear where
+those new changes must be applied.
+
+## Design Details
+
+### New data model
+
+Here we describe the new proposed data model in the form of Go structures, and non comprehensive sample CRDs.
+
+#### AddressPool
+
+The `AddressPool` is the entity responsible for the IP allocation part, regardless of how (and if) the ip is going to be advertised.
+
+Hence, we keep the same structure but remove the bgpAdvertisement section and the Protocol field.
+
+```go
+type addressPool struct {
+  Name              string
+  Addresses         []string
+  AvoidBuggyIPs     bool
+  AutoAssign        *bool
```

Comment:
Why a bool *pointer*?

Reply:
That's because the default value of AutoAssign is true (it's already like that in the current configuration), so we need a pointer to express "the user did not fill it" vs "the user asked for a false autoassign"

Reply:
Then why no pointer for `AvoidBuggyIPs`?

Reply:
Because avoidbuggyips default is false, so it's fine both if the user puts false or not specifies it.